Add copy link after token been sent for the user doesnt have email set up from design

peterhashair opened the pull request:
### Description:

Fixes: #19625

Add copy link after the token has been sent for the user who doesn't have an email set up from design. 

- As per the design, add a popup to resend invites. 
- As suggested, add an additional database field for the link token, and make 2 option works. 
- Add a new API endpoint.
- Add 2 new Icon, arrow left and arrow right
- As suggested, when over HTTPS using js `WriteText` otherwise `executeCommand('copy')`, there is still a security risk in `WriteText`, when failed, show the whole link.

This problem needs to update docs as well.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

Commit:
add copy link button and style

Commit:
add design copy link

add design copy link

Commit:
Merge branch '4.x-dev' into m19625

# Conflicts:
#	plugins/CoreHome/vue/dist/CoreHome.umd.js
#	plugins/CoreHome/vue/dist/CoreHome.umd.min.js

Commit:
Revert "add design copy link"

This reverts commit bb081bddb82691fb40c2a2d043e5058e7f9ca123.

Commit:
add link token

add link token

Commit:
update tests

update tests

peterhashair committed:
built vue files

Commit:
update xml tests

update xml tests

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

Commit:
update UI

update UI

peterhashair committed:
built vue files

Commit:
update ui

update ui

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

Commit:
Update UsersManagerTest.php

update tests

Commit:
update tests

update tests

Commit:
update tests

update tests

sgiehl commented on plugins/Morpheus/stylesheets/ui/_buttons.less:
```diff
@@ -73,4 +73,20 @@ input[type="submit"].btn,
+  position: absolute;
+  right: 20px;
+  top: 20px;
+  cursor: pointer;
```
That really doesn't look like a style that should be located in a global class. An absolute positioned element isn't anything that can be reused.

sgiehl commented on plugins/Morpheus/stylesheets/ui/_buttons.less:
```diff
@@ -73,4 +73,20 @@ input[type="submit"].btn,
+
+.btn-success {
+  border: 2px solid #5D9E52;
+  color: #5D9E52;
```
Where do you have that color from? We should try to reuse the common colors defined in https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Morpheus/stylesheets/base/colors.less, as they can't be overwritten by themes otherwise.

peterhashair replied:
that's out of date, I think

sgiehl commented on plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:
```diff
@@ -537,3 +548,11 @@ export default defineComponent({
 </script>
+
+<style scoped>
+.resend-link {
```
please avoid using scoped styles.

peterhashair replied:
that's out of date, I think

sgiehl commented on core/Db/Schema/Mysql.php:
```diff
@@ -55,6 +55,7 @@ public function getTablesCreateSql()
                           idchange_last_viewed TIMESTAMP NULL,
                           invited_by VARCHAR(100) NULL,
                           invite_token VARCHAR(191) NULL,
+                          invite_link_token VARCHAR(191) NULL,
```
Having two columns for an invite token, but only one for expiry doesn't make any sense at all. That means you extend the expiry of the token send by mail if you copy a new one in the UI.
I would still prefer the solution of only having one valid token. As long as it's explained in the UI, that generating or sending a new token invalidates the previous one, that should be fine.

peterhashair replied:
@sgiehl had a discussion with Justin, I guess we just go for an extra column with one expiration date for now.

Commit:
Update UsersManager_spec.js

update screenshots

Commit:
update ui

update ui

justinvelluppillai reviewed (changes requested):
I left a few screenshots and comments for you to get it conforming to spec in slack 👍🏽

Commit:
update css

update css

peterhashair committed:
built vue files

Commit:
remove password confirm for invite

remove password confirm for invite

peterhashair committed:
built vue files

Commit:
update css

update css

Commit:
convert vue scoped to less

convert vue scoped to less

Commit:
Update UsersManager.vue

fix vue error

Commit:
update vue tests

update vue tests

Commit:
Merge branch '4.x-dev' into m19625

# Conflicts:
#	plugins/UsersManager/stylesheets/usersManager.less

peterhashair committed:
built vue files

Commit:
update tests

update tests

Commit:
update ui

update ui

Commit:
update UserSetting token UI

update UserSetting token UI

Commit:
add new screen short copied success

add new screen short copied success

Commit:
revert failed UI

revert failed UI

sgiehl commented on plugins/UsersManager/API.php:
```diff
@@ -769,9 +771,10 @@ public function inviteUser($userLogin, $email, $initialIdSite = null, $expiryInD
-            $this->confirmCurrentUserPassword($passwordConfirmation);
-        }
+        //notes as design request this is removed.
+//        if (Common::getRequestVar('force_api_session', 0)) {
+//            $this->confirmCurrentUserPassword($passwordConfirmation);
+//        }
```
Then I guess we need to discuss the design again. Why should this be removed again, we actually just added it 🤷

peterhashair replied:
I know right, but the designer asks me to remove this.

sgiehl replied:
Sacrificing security for usability reasons isn't anything we should consider. Did you let @Javi-Ormaechea know, that this will make Matomo bit less secure? If that's a mandatory requirement you should at least let @tsteur confirm that.

tsteur replied:
From a security perspective, to invite a user via the UI we need to ask for a password confirmation for security purposes. 

I don't know where exactly we're asking for a password currently in the process. If it helps it could also look different. For example, that we only redirect to the "Invite user" screen after a password confirmation on a dedicated page. But then a user has only like 15 minutes to complete the action.

Javi-Ormaechea replied:
@tsteur I was not aware that this was a security issue but I'm curious to know more about why a password needs to be added here when the user has technically logged in already. As an experience it makes it less seamless, and probably unexpected too (_last bit is an assumption_).

tsteur replied:
@Javi-Ormaechea for sensitive actions we require a password confirmation as otherwise someone else could go onto your computer/laptop/tablet/... where you are logged in already and add a user on your behalf without your knowing. I don't know what the current screen looks like and if we ask for the password in the form. 

In some places we redirect the user for example first to a dedicated page for confirming the password and then redirect to the actual page. This password confirmation is then valid for I believe 15 minutes or so. Meaning there are options to change this.

peterhashair replied:
@tsteur I can't see the password confirmation remembered for 15 minutes.

sgiehl commented on plugins/Morpheus/stylesheets/base/icons.css:
```diff
@@ -403,3 +403,6 @@
 .icon-new_releases:before {
   content: "\e902";
 }
+.icon-arrow-left2:before {
```
new icons, if needed, should always be added to our demo page. Here https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Morpheus/templates/demo.twig#L621

Also I wonder if we directly should add the same arrow in the right direction.

peterhashair replied:
right direction added

sgiehl commented on plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:
```diff
@@ -503,7 +515,7 @@ export default defineComponent({
   computed: {
     formTitle() {
-      return this.isAdd ? translate('UsersManager_InviteNewUser') : translate('UsersManager_EditUser');
+      return this.isAdd ? translate('UsersManager_AddNewUser') : '';
```
Any reason for removing the `Edit user` in the headline?

peterhashair replied:
@sgiehl ah that was the design UI update see here. https://www.figma.com/file/ZAVoyUqPX7xUp858wOm45s/UI-Requests?node-id=57%3A2016

sgiehl commented on plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:
```diff
@@ -8,12 +8,18 @@
+      <div v-if="isAdd" class="col s12 m6 invite-notes">
+        <div class="form-help">
+                     <span v-html="$sanitize(
+                          translate('UsersManager_InviteSuccessNotification', [7]))"></span>
```
shouldn't that one use the config value, instead of a hard coded `7`?

sgiehl commented on plugins/Morpheus/stylesheets/ui/_buttons.less:
```diff
@@ -73,4 +73,11 @@ input[type="submit"].btn,
     padding: 0 2rem;
+}
+
+.btn-close {
```
This is a very global css class, but it contains very specific rules. The class is not reusable that way, or do you think every close button should be placed absolute? Might be better to move that where it is used and make the css selector more specific.

sgiehl commented on plugins/UsersManager/tests/System/expected/test___UsersManager.getUsersWithSiteAccess_3_admin_when_superuseraccess.xml:
```diff
@@ -19,6 +20,7 @@
 		
 		<invited_by />
 		
+		<invite_link_token />
```
The invite token are for sure nothing that should be returned in any API response. 
The are even stored hashed in database, so there is no use for having that at all. Those should be hiddne like the password and other fields. e.g. here https://github.com/matomo-org/matomo/blob/45058209303f51216367f07fde54f83a6928e1e6/plugins/UsersManager/Repository/UserRepository.php#L143-L153

peterhashair replied:
removed token field

sgiehl commented on plugins/UsersManager/vue/src/UserEditForm/UserEditForm.vue:
```diff
@@ -130,30 +138,29 @@
+              <span class="resend-link" @click="resendRequestedUser"
+                    v-html="$sanitize(translate('UsersManager_ResendInvite'))"></span>
+            </p>
+<!--            <PasswordConfirmation-->
+<!--              v-model="showPasswordConfirmationForInviteUser"-->
+<!--              @confirmed="inviteUser"-->
+<!--            >-->
+<!--              <h2 v-html="$sanitize(inviteUserTitle)"></h2>-->
+<!--              <p>{{ translate('UsersManager_ConfirmWithPassword') }}</p>-->
+<!--            </PasswordConfirmation>-->
```
Please ALWAYS avoid commenting out code that shouldn't be used anymore. It does not have any benefit keeping such stuff. If it needs to be reverted at some point, that can be easily done using git history. No need for polluting code.

sgiehl commented on plugins/UsersManager/lang/en.json:
```diff
@@ -191,18 +191,24 @@
+        "ExpiredInviteAutomaticallyRemoved": "Expired invites will be automatically deleted after %1$s days.",
+        "ResendInvite": "Resend invite",
+        "InvitationSent": "Invitation has been sent.",
+        "SendInvite": "Send Invite",
```
```suggestion
        "SendInvite": "Send invite",
```

sgiehl commented on plugins/UsersManager/lang/en.json:
```diff
@@ -191,18 +191,24 @@
+        "ResendInvite": "Resend invite",
+        "InvitationSent": "Invitation has been sent.",
+        "SendInvite": "Send Invite",
+        "CopyLink": "Copy Link",
```
```suggestion
        "CopyLink": "Copy link",
```

sgiehl commented on plugins/UsersManager/lang/en.json:
```diff
@@ -191,18 +191,24 @@
+        "InvitationSent": "Invitation has been sent.",
+        "SendInvite": "Send Invite",
+        "CopyLink": "Copy Link",
+        "LinkCopied": "Link Copied",
```
```suggestion
        "LinkCopied": "Link copied",
```

sgiehl commented on plugins/UsersManager/lang/en.json:
```diff
@@ -191,18 +191,24 @@
+        "SendInvite": "Send Invite",
+        "CopyLink": "Copy Link",
+        "LinkCopied": "Link Copied",
+        "InviteLinkValidDays": "Invite links are valid for %1$s-days.",
```
```suggestion
        "InviteLinkValidDays": "Invite links are valid for %1$s days.",
```

sgiehl commented on plugins/CoreHome/stylesheets/coreHome.less:
```diff
@@ -9,7 +9,7 @@
 .card h2 {
     &.card-title,
     .card-title {
-        padding: 0 !important;
```
Are you sure this won't have any side effects in core or for any plugin?

peterhashair replied:
I believe this is no effect from I can aware.

peterhashair replied:
revert that one

sgiehl reviewed (changes requested):
Added a couple of comments.
And to repeat what I already mentioned on the previous PRs:
I don't think that the usability of this new feature is very good. Even I, as someone quite familiar with software having broken usability, would not expect that I first need to send an invite by mail before being able to generate a new link for copying it.
It would be so much simpler and smarter to initially provide two buttons, to either invite by mail or copying the link. That would also avoid any sense of having two separate tokens.

Also I'm not sure if only providing a button to copy the link to clipboard is the optimal solution. Some people might prefer having a textbox, where they can copy the link from.

But that's up to @justinvelluppillai and @Javi-Ormaechea, to decide if we should implement such improvements...

Commit:
Update plugins/UsersManager/lang/en.json

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Update plugins/UsersManager/lang/en.json

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Update plugins/UsersManager/lang/en.json

Co-authored-by: Stefan Giehl <stefan@matomo.org>

Commit:
Update plugins/UsersManager/lang/en.json

Co-authored-by: Stefan Giehl <stefan@matomo.org>

peterhashair commented:
@justinvelluppillai any suggestion on that one? Should I go back to @Javi-Ormaechea?

Commit:
remove DB field and convert the copy link

remove DB field and convert the copy link

Commit:
update error

update error

Commit:
update notification

update notification

peterhashair committed:
built vue files

Commit:
revert database changes

revert database changes

Commit:
update api and tests

update api and tests

peterhashair committed:
built vue files

Commit:
update padding css

update padding css

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

Commit:
move global button css to modal

move global button css to modal

Commit:
update modal to global

update modal to global

Commit:
remove invite token

remove invite token

peterhashair commented on plugins/CoreHome/vue/src/Notification/Notification.ts:
```diff
@@ -82,6 +82,7 @@ interface Notification {
    * If true, the notification will be displayed before others currently displayed.
    */
   prepend?: boolean;
+
```
```suggestion
```

Commit:
Update plugins/CoreHome/vue/src/Notification/Notification.ts

Commit:
update fonts and screenshots

update fonts and screenshots

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

Commit:
update fonts

update fonts

Commit:
Merge branch '4.x-dev' into m19625

Commit:
exclude woff2 file on release check

exclude woff2 file on release check

peterhashair commented:
Notes: I believe once that merge we need to update the docs here https://matomo.org/faq/general/manage-users/

peterhashair commented on plugins/UsersManager/API.php:
```diff
@@ -769,9 +771,10 @@ public function inviteUser($userLogin, $email, $initialIdSite = null, $expiryInD
-        if (Common::getRequestVar('force_api_session', 0)) {
-            $this->confirmCurrentUserPassword($passwordConfirmation);
-        }
+        //notes as design request this is removed.
+//        if (Common::getRequestVar('force_api_session', 0)) {
+//            $this->confirmCurrentUserPassword($passwordConfirmation);
+//        }
```
```suggestion
       if (Common::getRequestVar('force_api_session', 0)) {
           $this->confirmCurrentUserPassword($passwordConfirmation);
       }
```

Commit:
Update plugins/UsersManager/API.php

Commit:
update tests

update tests

Commit:
remove duplicated

remove duplicated

Commit:
update confirm box

update confirm box

peterhashair committed:
built vue files

Commit:
tidy up some file

tidy up some file

sgiehl commented on plugins/Morpheus/stylesheets/base/icons.css:
```diff
@@ -403,3 +403,9 @@
 .icon-new_releases:before {
   content: "\e902";
 }
+.icon-arrow-left2:before {
```
for consistency we should maybe name the icons with a `-`.
e.g. `icon-arrow-left-2`, like you used it in demo.twig, which currently doesn't work as they have another name..

sgiehl replied:
for consistency we should maybe name the icons with a `-`.
e.g. `icon-arrow-left-2`, like you used it in demo.twig, which currently doesn't work as they have another name..

sgiehl commented on plugins/UsersManager/API.php:
```diff
@@ -1522,13 +1524,12 @@ private function getDeviceDescription()
      * @throws NoAccessException
      */
-    public function resendInvite($userLogin, $expiryInDays = 7)
+    public function generateInviteLink($userLogin, $mail = false, $expiryInDays = 7)
```
It's actually a bad solution to have a method that is able to do different things. It might not be clear for a user that the API method is either able to send a mail or returns an invite link. The method name actually only indicates the latter one. Also when changing API method, please ensure to also update the docs blocks accordingly.

sgiehl commented on plugins/UsersManager/API.php:
```diff
@@ -1545,13 +1546,31 @@ public function resendInvite($userLogin, $expiryInDays = 7)
-         */
-        Piwik::postEvent('UsersManager.inviteUser.resendInvite', [$userLogin, $user['email']]);
+        } else {
+            $token = $this->userRepository->generateInviteToken($userLogin, (int)$expiryInDays);
```
An API method to return a invitation link, should for security reasons require a password when triggered through API session.

sgiehl commented on plugins/UsersManager/lang/en.json:
```diff
@@ -191,18 +191,24 @@
         "InviteSuccessNotification": "The invited user will receive an email to accept the invite. This invite is valid for %1$s days. You can also resend and delete the invite on the user management page.",
-        "ResendInviteConfirmSingle": "Are you sure you want to resend invitation to %1$s?",
-        "InviteConfirm": "Are you sure you want to send an invitation to %1$s?",
+        "InviteConfirm": "You can resend the invitation by copying the invite link and sharing directly with %1$s, or resend an invite email to %2$s.",
```
Changing the contents of a translations key is only suitable if it's content doesn't change too much. In that case using a new name might be better, as existing translations don't have the second placeholder.

sgiehl commented on plugins/UsersManager/API.php:
```diff
@@ -1545,13 +1546,31 @@ public function resendInvite($userLogin, $expiryInDays = 7)
+             *
+             * @param string $userLogin The new user's login.
+             */
+            Piwik::postEvent('UsersManager.inviteUser.generateInviteLinkToken', [$userLogin, $user['email']]);
```
New events should always be mentioned in the developer changelog.

sgiehl commented on plugins/UsersManager/vue/src/UsersManager/UsersManager.vue:
```diff
@@ -305,17 +354,29 @@ export default defineComponent({
       });
     },
+    generateInviteLink(user: User) {
+      AjaxHelper.fetch<{ value: string }>(
```
I don't think it makes much sense to send a request every time the button is clicked. Some people e.g. tend to do double click on buttons. That would currently cause two simultaneous requests being sent out. In theory it could even happen that the request being resolved last, was not the one request, that updated the token last. So it could happen that you could hold a link on clipboard that is not even valid anymore.
I would suggest to disable the button, as long as a request is running. And maybe even avoid multiple requests for the same user. If the first request returned a link it should be enough to hold that link in memory every time the button is clicked - until the modal is closed.

sgiehl commented on core/Updates/4.12.0-b2.php:
```diff
@@ -38,11 +38,12 @@ public function __construct(MigrationFactory $factory)
     public function getMigrations(Updater $updater)
     {
         $uniqueIndex = $this->migration->db->addUniqueKey('user', 'email', 'uniq_email');
+        $linkToken = $this->migration->db->addColumns('user', ['invite_link_token' => 'VARCHAR(191) DEFAULT null']);
```
Please never extend any update scripts. If you have to add changes, always increase the version in `Version.php` and use this new version for your update script. Otherwise this can cause problems if the update script was already executed for any install for any reason.

sgiehl reviewed (changes requested):
Left a couple of comments for possible improvements. Besides that I'll summarize the concerns I already mentioned in the PR(s) before:

The usability of that new "feature" really feels broken. I can't see a useful reason why you first need to send out an invitation email before being able to copy an invitation link.
Even though it might have been in the UX designs, it doesn't make sense.
Even more: To fulfill this UX requirement we now even store two different invitation tokens, so that the token sent out with the first mail is not directly invalidated when directly using the copy link feature afterwards.

That way of storing two token actually also contains a "small" security risk, as we still only have one expiry time. So sending out an email or copying a new link would magically extend the expiration of the other token. Which might be unexpected and less secure.

In terms of security we currently also only ask for a password when the user is initially invited. That makes it impossible for someone gaining access to an admin account (without knowing the password) to create a new user. But the feature to copy a new invite link currently does not ask for a password. So if there is an invited user, that has not yet accepted, an attacker would easily be able to take over that account by simply copying the invite link, without anyone noticing.


----

What I would have expected from this new feature is an implementation like:

- The form to add / invite a new user would instead of only asking for a password, show a modal similar to the resend invitation modal in this PR, but with an additional input for the password.
- This would allow to directly choose between sending an email or copying the link. Clicking one of the buttons would create the user and either send out an email or return the link to provide it to the user for copying it.
- Having the direct choice would also make the additional database field obsolete, as one token would be enough at a time.
- Clicking the copy link button should imho also show the link in a textarea or similar. There might be browsers / users where copying something to clipboard might be restricted for some reason. That would currently make it impossible to actually get the link.
- The modal to resend an invitation would actually look the same, but it would also include a password field for security reasons.

Commit:
update icon and update version number

update icon and update version number

justinvelluppillai commented:
@tsteur do you have any thoughts as security person on https://github.com/matomo-org/matomo/pull/19707#pullrequestreview-1107407965? We have included all security and UX/UI suggestions received so far, be good to know if there are any raised from @sgiehl that you'd like added here also.

justinvelluppillai commented:
Meantime @peterhashair please proceed to address @sgiehl's other review comments, they all seem important.

Commit:
update change log and translation

update change log and translation

Commit:
update button trigger generate link twice

update button trigger generate link twice

Commit:
Merge branch '4.x-dev' into m19625

# Conflicts:
#	core/Updates/4.12.0-b3.php

Commit:
add updates

add updates

tsteur commented:
From what I see the password confirmation check is still there so this should be all good 👍

> That way of storing two token actually also contains a "small" security risk, as we still only have one expiry time. So sending out an email or copying a new link would magically extend the expiration of the other token. Which might be unexpected and less secure.

It's tricky. 

* We could not extend the expire time when clicking `copy link`. This could be fine as long as we show information for how long that link is valid after copying it. It's actually currently also not clear that clicking "Resent invite" will extend the 7 days. At least the wording doesn't suggest that clearly. 
* Or we still extend the time but then make it more clear that any action will extend the allowed invitation time. We just need to be clear about it at least. 

> In terms of security we currently also only ask for a password when the user is initially invited. That makes it impossible for someone gaining access to an admin account (without knowing the password) to create a new user. But the feature to copy a new invite link currently does not ask for a password. So if there is an invited user, that has not yet accepted, an attacker would easily be able to take over that account by simply copying the invite link, without anyone noticing.

It's a hard one. I would say if we're in doubt / unsure we should decide pro security. Especially since the action may be executed rarely and it's not a standard flow I would recommend asking for password confirmation to copy the link. 

fyi getting this error in the UI (not sure if it's known)

<img width="1063" alt="image" src="https://user-images.githubusercontent.com/273120/190290637-cd00c6fa-9c2c-4234-aae2-ad47892013aa.png">

Unrelated: Not sure if it's important but the header still says "Add new user" when you're actually inviting a new user. It could give the indication that a user is added directly maybe. Not needed to be changed though, just an FYI that this is maybe not quite clear. Especially for existing people where they maybe don't read the buttons as they've been using them many times before.
<img width="573" alt="image" src="https://user-images.githubusercontent.com/273120/190290922-38121aef-8190-4318-a535-beef37f45777.png">

I then couldn't find the feature to get the link but eventually found it by clicking on "Resend invite". I don't know what's been discussed so far and don't want to make anything more complicated. Was just thinking it could be useful if it said `Not received the invite? Resend invite or copy activation link` or something. Just a thought.

Commit:
add missing param

add missing param

Commit:
update missing comma

update missing comma

Commit:
remove un use variable

remove un use variable

Commit:
update tests

update tests

peterhashair committed:
built vue files

Commit:
add password confirm for the invite action and add popup notes

add password confirm for the invite action and add popup notes

peterhashair commented:
Update the PR for the following, just random thoughts unrelated. I would suggest, that we actually remove copy link, it may cause other issues later on (like copy function in js won't work). instead, we could give 2 actions which are `create a user` and `invite a user. `create a user` goes to the old way, admin/super admin enters the password for the client. 'Invite User' only available when a client has `email_enabled` turned on.  



- Updated the resend and copy link, they both required a passwords confirmation to trigger the action, if the password was wrong, the popup will remain open until the correct password is entered

- add invite action notes, `any action will extend the invite link to x days`

Commit:
update vue error

update vue error

Commit:
revert 4.120-b2 update to b4

revert 4.120-b2 update to b4

Commit:
update vue error

update vue

peterhashair committed:
built vue files

Commit:
update typescript error

update typescript error

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

Commit:
update type script

update type script

Commit:
update vue error

update vue error

peterhashair committed:
built vue files

Commit:
update tests

update tests

Commit:
update screenshots

update screenshots

Commit:
update tests

update tests

Commit:
update tests and wording on the popup

update tests and wording on the popup

peterhashair committed:
built vue files

Commit:
adding clipboard copy error in safari etc

adding clipboard copy error in safari etc

peterhashair committed:
built vue files

Commit:
update double error message

update double error message

Commit:
update wording

update wording

peterhashair committed:
built vue files

Commit:
hot fix give default period (#19724)

* hot fix give default period

hot fix give default period

* update to user default period

update to user default period

* add default date param in templates

add default date param in templates

Commit:
update tests prevent random fail

update tests prevent random fail

peterhashair committed:
built vue files

Commit:
update ui

update ui

Commit:
Merge branch 'm19625' of github.com:matomo-org/matomo into m19625

peterhashair commented:
As @Javi-Ormaechea requested, update the wording. It seems like our UI tester deny clipboard access, so leave the screen as that copy denied, and in order to prevent random failure with the random token, hide the token link. 

I believe there is a security risk that another site can access the clipboard which accesses the token link, the chance is rare. I think the clipboard is not meant to copy passwords etc. 

I would suggest, giving up this PR, recreating an Issue that uses `invite user` for our cloud and those who have email setup,  and `create user` as it was (admin enter password) for the user who doesn't have an email setup.

sgiehl committed:
Improve no data check in html visualization (#19740)

Commit:
add command copy over http

add command copy over http

peterhashair committed:
built vue files

snake14 committed:
Adding the abililty to exclude specific sites from the site selector (#19738)

* Adding the abililty to exclude specific sites from the Vue site selector component.

* Adding more time to make sure that the select is loaded before trying to search.

* Changed things to filter the site on the server side.

* Added a few new test cases.

* Increasing amount of time for test.

* Updated the omnifixture.

* Fixing logic error in view.

* Removed collation from omnifixture since it was causing issues.

* Switching omnifixture back to 4.x-dev version.

* Making just the bare essential changes to make local test fixtures build correctly.

* Updating screenshot of API list.

* revert OmniFixture-dump to 4.x-dev

revert OmniFixture-dump to 4.x-dev

Co-authored-by: Peter <peter@innocraft.com>

Commit:
Merge branch '4.x-dev' into m19625

# Conflicts:
#	tests/UI/expected-screenshots/UIIntegrationTest_api_listing.png

Commit:
merge 4.x and update copied success ui tests

merge 4.x and update copied success ui tests

Commit:
Merge branch 'next_release' into m19625

Commit:
rebase to next release

rebase to next release

Commit:
rebase to next release

rebase to next release

justinvelluppillai reviewed (changes requested):
In safari the copied link shows up in the header when I click "Copy Link" (see screenshot):

<img width="886" alt="Screen Shot 2022-09-30 at 12 23 26 PM" src="https://user-images.githubusercontent.com/1127121/193159059-93c2ca78-6014-479d-ab68-9a17a629416a.png">

Also when I click Copy Link in the modal in chrome, the modal remains open but the Copy Link and Resend Invite buttons now do nothing when clicked.

Commit:
update safrai and enable copy success and copy again

update safrai and enable copy success and copy again

peterhashair commented:
@justinvelluppillai should be fixed now.

Commit:
fix Safrai showing random token

fix Safrai showing random token

peterhashair committed:
built vue files

bx80 commented:
This is my first look at this PR, it does work :+1: and I'm aware there have already been a lot of points raised in the discussion above, I'll try to avoid rehashing previously covered comments.

a) If I want to create a new user and send them their invite link on slack, then I'd probably end up doing this:

1) Click 'Invite a new user' (goes to a screen labeled 'Add a new user')
2) Click 'Invite User'.
3) Enter my password, then 'Invitation is sent' is shown.
4) Go back to the user list.
5) Search for the newly invited user (what was their email again? Type it in...)
6) Click on an envelope row icon ("but I don't want to email, I want to copy...")
7) Click 'Copy link' on the popup.
8) Enter my password again (second time in 30 seconds :slightly_frowning_face: )
9) Finally the invite link is in the clipboard.

I did this a couple of times before I realized that the 'Resend Invite' link shown in step 3) would allow me to jump to step 7). So although I saw 'Resend Invite' link, it didn't register because I want to *copy* the link. I never wanted to *send* an invite in the first place, so wasn't intuitive. 

**Could we either rename this to 'Copy Invite Link' or duplicate the link so we have 'Resend invite' and 'Copy invite link'?** 

b) Once the 'copy link' button is used the buttons are disabled but it isn't obvious that they no longer work. If I accidentally copied something else into my clipboard and lost the copied link then clicking the 'Copy link' button again has no effect. Wouldn't it be better to hide the buttons altogether and/or replace them with a close button?

c) Minor cosmetic: The resend invite popup box seems to have a different font size from the normal popups? The password popup font is much smaller, it looks like part of a different application.

d) Minor point: It might be slightly clearer to say 'Link copied to the clipboard' rather than just 'Link copied'.

e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note it somewhere else for improvement as we add password prompts to more places.

peterhashair commented:
@bx80 I agree.
> e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note this somewhere else for improvement, since we're adding password prompts to more places in the UI and risk increasing user frustration.

I guess this has come to the design question. @Javi-Ormaechea 
> Minor point: It might be slightly clearer to say 'Link copied to the clipboard' rather than just 'Link copied'.

Commit:
Revert "Adding the abililty to exclude specific sites from the site selector (#19738)"

This reverts commit a53a62892f540ea683eb56949bb7d2172731c2cc.

bx80 commented:
> >    e) Having to enter the password twice in the space of 10 seconds isn't great, probably beyond the scope of this PR, but maybe we note this somewhere else for improvement, since we're adding password prompts to more places in the UI and risk increasing user frustration.

>I guess this has come to the design question.

We could possibly look at a system level solution - same as `sudo` won't prompt for your password over and over if you entered it in the last 5 minutes. At the moment being tasked with adding 10 new users and copying their invites would make the admin type their password 20 times in under 5 minutes, it might be secure but it's the sort of UX that causes the admin to change their password to `12345` out of frustration :slightly_smiling_face:  I can log a RFC issue on this if you like?

justinvelluppillai commented:
@bx80 yes it'd be good to maybe create a separate issue for that. I agree it wasn't fun getting links or sending invites during testing 👍🏽

Commit:
update fonts and text

update fonts and text

peterhashair committed:
built vue files

Commit:
update wording

update wording

peterhashair committed:
built vue files

Commit:
Updated UI test screenshots

bx80 reviewed (approved):
I've spent some time doing functional testing, apart from the multiple password entry annoyance which we can address elsewhere, it works as expected and allows an admin to get an invite link copied to the clipboard fairly quickly after creating the initial invite. I can't see any outstanding issues with the code.

I've pushed a couple of  UI test screenshot updates, once the test builds complete without issues then I think this is ready to merge :+1:

* [x] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [x] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [x] It's usable. [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [x] Can't see any issues [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [x] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [x] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [x] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [x] Updated [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [x] Updated  [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [x] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [x] Existing documentation updated if needed

sgiehl commented:
@peterhashair There is no need to recreate pull requests when you want to target another branch. You can simply change the target branch of a PR on GitHub. Also it is important to **rebase** your changes to the new target branch.
This PR actually included some other PRs that had been merged to 4.x-dev. It doesn't hurt that those changes are now included in the release. But it's important to keep any eye on the changes your PRs include.
In general you should avoid closing PRs in order to create a new one afterwards. Unless the changes included in the PR are completely different it would be good to always use the same PR till it is merged. This ensures that all comments that were made can be directly seen. By closing and recreating a PR you actually kind of loose all comments done before, as no one will search if there might have been previous PRs.

peterhashair commented:
@sgiehl yes, sorry. I confused myself there, after I recreate the PR, and release I can rebase the PR. Will not do that again.
